### PR TITLE
[rllib] Added vanilla policy gradient

### PR DESCRIPTION
**Pull request by alvkao58**
Implemented vanilla policy gradient in RLLib.

### CODE REVIEW

**Comment by richardliaw** [comment]
ok to test

File: python/ray/rllib/pg/pg.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    # Use LSTM model - only applicable for image states
+    "observation_filter": "NoFilter",
+    # Which reward filter to apply to the reward
+    "reward_filter": "NoFilter",
+    # Discount factor of MDP
+    "gamma": 0.99,
+    # GAE(gamma) parameter
+    "lambda": 1.0,
+    # Max global norm for each gradient calculated by worker
+    "grad_clip": 10.0,
```
Comment (ericl) [inline python/ray/rllib/pg/pg.py]: Shall we drop the gae / filter / clipping stuff to simplify the example?

Comment (richardliaw) [reply python/ray/rllib/pg/pg.py]: yes I think that's a really good idea; @alvkao58 what are your thoughts?

Comment (alvkao58) [reply python/ray/rllib/pg/pg.py]: Sure, that sounds good, I think I'll also drop the stuff relevant to the value function for sake of simplicity.

**Comment by ericl** [comment]
Could you add a test entry in https://github.com/ray-project/ray/blob/master/python/ray/rllib/test/test_supported_spaces.py ?

File: python/ray/rllib/pg/pg.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        # Number of gradients applied for each `train` step
+        "grads_per_step": 10,
+    },
+    "model_options": {},
+    # Arguments to pass to the env creator
+    "env_config": {},
+}
+
+
+class PGAgent(Agent):
```
Comment (ericl) [inline python/ray/rllib/pg/pg.py]: """Simple policy gradient agent.

This is an example agent to show how to implement algorithms in RLlib. In most cases, you will probably want
to use the PPO agent instead."""

Comment (ericl) [reply python/ray/rllib/pg/pg.py]: Could you also switch the link in http://ray.readthedocs.io/en/latest/rllib-dev.html from A3CEvaluator to PGAgent instead? (rllib-dev.rst file under doc/source)

Comment (alvkao58) [reply python/ray/rllib/pg/pg.py]: Done; I switched the link to be to PGEvaluator instead of PGAgent though (since it's supposed to be an example of an evaluator class).

File: python/ray/rllib/pg/pg.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    "model_options": {"fcnet_hiddens": [32, 32]},
+    # Arguments to pass to the env creator
+    "env_config": {},
+}
+
+
+"""Simple policy gradient agent.
+
+This is an example agent to show how to implement algorithms in RLlib. In most cases, you will probably want
+to use the PPO agent instead."""
```
Comment (ericl) [inline python/ray/rllib/pg/pg.py]: This should go inside the class def (class doc)

File: doc/source/rllib-dev.rst [hunk 4]
```diff
@@ -10,7 +10,7 @@ Recipe for an RLlib algorithm
 
 Here are the steps for implementing a new algorithm in RLlib:
 
-1. Define an algorithm-specific `Policy evaluator class <#policy-evaluators-and-optimizers>`__ (the core of the algorithm). Evaluators encapsulate framework-specific components such as the policy and loss functions. For an example, see the `A3C Evaluator implementation <https://github.com/ray-project/ray/blob/master/python/ray/rllib/a3c/a3c_evaluator.py>`__.
+1. Define an algorithm-specific `Policy evaluator class <#policy-evaluators-and-optimizers>`__ (the core of the algorithm). Evaluators encapsulate framework-specific components such as the policy and loss functions. For an example, see the `PGEvaluator implementation <https://github.com/ray-project/ray/blob/master/python/ray/rllib/pg/pg_evaluator.py>`__.
```
Comment (ericl) [inline doc/source/rllib-dev.rst]: see the simple policy gradient evaluator example

File: python/ray/rllib/pg/pg_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+from ray.rllib.pg.policy import PGPolicy
+from ray.rllib.utils.filter import NoFilter
+from ray.rllib.utils.process_rollout import process_rollout
+from ray.rllib.utils.sampler import SyncSampler
+
+# Evaluator for vanilla policy gradient
+
+
+class PGEvaluator(Evaluator):
+    """Actor for simple policy gradient."""
```
Comment (ericl) [inline python/ray/rllib/pg/pg_evaluator.py]: Evaluator for

File: python/ray/rllib/pg/pg.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        episode_rewards = []
+        episode_lengths = []
+        metric_lists = [a.get_completed_rollout_metrics.remote()
+                        for a in self.remote_evaluators]
+        for metrics in metric_lists:
+            for episode in ray.get(metrics):
+                episode_lengths.append(episode.episode_length)
+                episode_rewards.append(episode.episode_reward)
+        avg_reward = (
+            np.mean(episode_rewards) if episode_rewards else float('nan'))
```
Comment (richardliaw) [inline python/ray/rllib/pg/pg.py]: hm I have to double check this but perhaps this `if-else` doesn't actually need to be here

File: python/ray/rllib/pg/policy.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                                          tf.get_variable_scope().name)
+
+    def _setup_loss(self, action_space):
+        self.ac = ModelCatalog.get_action_placeholder(action_space)
+        self.adv = tf.placeholder(tf.float32, [None], name="adv")
+
+        log_prob = self.dist.logp(self.ac)
+
+        # policy loss
+        self.loss = -tf.reduce_sum(log_prob * self.adv)
```
Comment (richardliaw) [inline python/ray/rllib/pg/policy.py]: consider `reduce_mean`

File: python/ray/rllib/pg/policy.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        self.sess.run(self._apply_gradients, feed_dict=feed_dict)
+
+    def get_weights(self):
+        return self.variables.get_weights()
+
+    def set_weights(self, weights):
+        self.variables.set_weights(weights)
+
+    def compute(self, ob, *args):
+        action = self.sess.run([self.sample], {self.x: [ob]})
```
Comment (richardliaw) [inline python/ray/rllib/pg/policy.py]: we don't need a bracket around `self.sample` right? this would simplify the indenting below.

File: python/ray/rllib/pg/policy.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            self.ac: samples["actions"],
+            self.adv: samples["advantages"],
+        }
+        self.grads = [g for g in self.grads if g is not None]
+        self.local_steps += 1
+        grad = self.sess.run(self.grads, feed_dict=feed_dict)
+        return grad, info
+
+    def apply_gradients(self, grads):
+        feed_dict = {self.grads[i]: grads[i]
```
Comment (richardliaw) [inline python/ray/rllib/pg/policy.py]: consider `feed_dict = dict(zip(self.grads, grads))`

File: python/ray/rllib/pg/policy.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+class PGPolicy():
+
+    other_output = []
+    is_recurrent = False
+
+    def __init__(self, registry, ob_space, ac_space, config):
+        name = "local"
+        self.registry = registry
+        self.local_steps = 0
```
Comment (richardliaw) [inline python/ray/rllib/pg/policy.py]: this isn't really used right? i would be ok with dropping it

File: python/ray/rllib/pg/policy.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        self.config = config
+        worker_device = "/job:localhost/replica:0/task:0/cpu:0"
+        self.g = tf.Graph()
+        with self.g.as_default(), tf.device(worker_device):
```
Comment (richardliaw) [inline python/ray/rllib/pg/policy.py]: I'm not so sure about the use of `worker_device` here - I can take a look into this, and chances are we don't actually need it.

File: python/ray/rllib/pg/pg.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        result = TrainingResult(
+            episode_reward_mean=avg_reward,
+            episode_len_mean=avg_length,
+            timesteps_this_iter=timesteps,
+            info={})
+
+        return result
+
+    def compute_action(self, observation):
+        obs = self.local_evaluator.obs_filter(observation, update=False)
```
Comment (richardliaw) [inline python/ray/rllib/pg/pg.py]: filters are dropped right?

File: python/ray/rllib/pg/pg_evaluator.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class PGEvaluator(Evaluator):
+    """Evaluator for simple policy gradient."""
+
+    def __init__(self, registry, env_creator, config):
+        self.env = ModelCatalog.get_preprocessor_as_wrapper(
+                registry, env_creator(config["env_config"]), config["model"])
+        self.config = config
+        self.registry = registry
```
Comment (richardliaw) [inline python/ray/rllib/pg/pg_evaluator.py]: we don't need this right?

**Review by richardliaw — COMMENTED** [review]
Overall this is solid! Left a few comments. I think we should have a `tuned_example` (like on CartPole` or something.

File: python/ray/rllib/pg/policy.py (hunk 11 above)
Comment (ericl) [reply python/ray/rllib/pg/policy.py]: Yeah you don't want to do this, TF will pick devices better by itself.

Comment (ericl) [reply python/ray/rllib/pg/policy.py]: You also shouldn't need the tf.Graph thing, agents already set up a default graph. In fact this might mess up checkpointing.

**Review by ericl — COMMENTED** [review]
Probably should also add this to the test checkpoint restore test in Rllib/test

File: python/ray/rllib/pg/policy.py (hunk 11 above)
Comment (richardliaw) [reply python/ray/rllib/pg/policy.py]: yeah, this is copied from A3C so I should fix it there too

File: python/ray/rllib/pg/pg_evaluator.py (hunk 13 above)
Comment (alvkao58) [reply python/ray/rllib/pg/pg_evaluator.py]: I think we still need self.config for the discount factor.

Comment (richardliaw) [reply python/ray/rllib/pg/pg_evaluator.py]: ah this comment was in reference for `self.registry` but has since been removed.

**Comment by richardliaw** [comment]
ok I think the last todos are 
1. add jenkins tests to make sure it works for some atari and classical control domain
2. make sure it hits 200 in CartPole-v0 and learns something in Acrobot-v0.

File: python/ray/rllib/pg/pg.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    # Discount factor of MDP
+    "gamma": 0.99,
+    # Learning rate
+    "lr": 0.0005,
+    # Arguments to pass to the rllib optimizer
+    "optimizer": {
+        # Number of gradients applied for each `train` step
+        "grads_per_step": 1,
+    },
+    "model_options": {"fcnet_hiddens": [32, 32]},
```
Comment (richardliaw) [inline python/ray/rllib/pg/pg.py]: Let's call this `model` to keep consistency with A3C, PPO, etc

File: python/ray/rllib/pg/pg_evaluator.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+from __future__ import print_function
+
+import ray
+from ray.rllib.optimizers import Evaluator
+from ray.rllib.pg.policy import PGPolicy
+from ray.rllib.utils.filter import NoFilter
+from ray.rllib.utils.process_rollout import process_rollout
+from ray.rllib.utils.sampler import SyncSampler
+
+# Evaluator for vanilla policy gradient
```
Comment (richardliaw) [inline python/ray/rllib/pg/pg_evaluator.py]: Let's remove this

**Review by richardliaw — CHANGES_REQUESTED** [review]
A few more minor nits. Note - this is also failing for Pong-v0, so it'd be nice to make sure that's fixed before this is merged.

For testing, can you add the following to `test/jenkins_tests/run_multi_node_tests.sh`:

```
docker run --rm --shm-size=10G --memory=10G $DOCKER_SHA \
    python /ray/python/ray/rllib/train.py \
    --env CartPole-v0 \
    --run PG \
    --stop '{"training_iteration": 2}' \
    --config '{"batch_size": 500, "num_workers": 1}'

docker run --rm --shm-size=10G --memory=10G $DOCKER_SHA \
    python /ray/python/ray/rllib/train.py \
    --env Pong-v0 \
    --run PG \
    --stop '{"training_iteration": 2}' \
    --config '{"batch_size": 500, "num_workers": 1}'

docker run --rm --shm-size=10G --memory=10G $DOCKER_SHA \
    python /ray/python/ray/rllib/train.py \
    --env FrozenLake-v0 \
    --run PG \
    --stop '{"training_iteration": 2}' \
    --config '{"batch_size": 500, "num_workers": 1}'
```

**Review by richardliaw — APPROVED** [review]

**Comment by richardliaw** [comment]
Looks good! I'll merge right after this last rerun passes.
